### ISSUE TITLE
Invalid platform version "" specified: upgrade to version 2.0.30 fails

### ISSUE THREAD

**Issue by hkockerbeck** [kimai/kimai#4246]
### Describe the issue

When I try to upgrade from Kimai 2.0.29 to version 2.0.30, clearing the cache fails:

```
<redacted>$ git checkout 2.0.30
Vorherige Position von HEAD war b8315323 Release 2.0.29 (#4178)
HEAD ist jetzt bei adc07799 Release 2.0.30 (#4225)
```
```
<redacted>$ php -d allow_url_fopen=On ~/composer.phar install --no-dev --optimize-autoloader
Installing dependencies from lock file
Verifying lock file contents can be installed on current platform.
Package operations: 0 installs, 57 updates, 0 removals
  - Upgrading symfony/flex (v2.3.1 => v2.3.3): Extracting archive
  - Upgrading symfony/runtime (v6.2.8 => v6.2.13): Extracting archive
  - Upgrading symfony/routing (v6.2.8 => v6.2.13): Extracting archive
  - Upgrading symfony/http-foundation (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/event-dispatcher (v6.2.8 => v6.2.13): Extracting archive
  - Upgrading symfony/var-dumper (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/error-handler (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/http-kernel (v6.2.11 => v6.2.14): Extracting archive
  - Upgrading symfony/finder (v6.2.7 => v6.2.14): Extracting archive
  - Upgrading symfony/filesystem (v6.2.10 => v6.2.12): Extracting archive
  - Upgrading symfony/var-exporter (v6.2.10 => v6.2.13): Extracting archive
  - Upgrading symfony/dependency-injection (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/config (v6.2.7 => v6.2.13): Extracting archive
  - Upgrading symfony/cache (v6.2.10 => v6.2.13): Extracting archive
  - Upgrading symfony/framework-bundle (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/string (v6.2.8 => v6.2.13): Extracting archive
  - Upgrading symfony/console (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading doctrine/dbal (3.6.3 => 3.6.5): Extracting archive
  - Upgrading symfony/doctrine-bridge (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading doctrine/doctrine-bundle (2.10.0 => 2.10.2): Extracting archive
  - Upgrading doctrine/inflector (2.0.6 => 2.0.8): Extracting archive
  - Upgrading doctrine/collections (2.1.2 => 2.1.3): Extracting archive
  - Upgrading doctrine/orm (2.15.2 => 2.16.1): Extracting archive
  - Upgrading gedmo/doctrine-extensions (v3.11.1 => v3.12.0): Extracting archive
  - Upgrading phpstan/phpdoc-parser (1.22.0 => 1.23.1): Extracting archive
  - Upgrading jms/serializer (3.25.0 => 3.27.0): Extracting archive
  - Upgrading twig/twig (v3.6.1 => v3.7.0): Extracting archive
  - Upgrading symfony/twig-bridge (v6.2.8 => v6.2.13): Extracting archive
  - Upgrading league/csv (9.9.0 => 9.10.0): Extracting archive
  - Upgrading setasign/fpdi (v2.3.7 => v2.4.1): Extracting archive
  - Upgrading zircote/swagger-php (4.7.10 => 4.7.11): Extracting archive
  - Upgrading phpdocumentor/type-resolver (1.7.2 => 1.7.3): Extracting archive
  - Upgrading nelmio/api-doc-bundle (v4.11.1 => v4.12.0): Extracting archive
  - Upgrading maennchen/zipstream-php (3.0.2 => 3.1.0): Extracting archive
  - Upgrading symfony/property-access (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/security-http (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/security-csrf (v6.2.7 => v6.2.13): Extracting archive
  - Upgrading symfony/security-bundle (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading scheb/2fa-bundle (v6.8.0 => v6.9.0): Extracting archive
  - Upgrading scheb/2fa-backup-code (v6.8.0 => v6.9.0): Extracting archive
  - Upgrading scheb/2fa-totp (v6.8.0 => v6.9.0): Extracting archive
  - Upgrading symfony/css-selector (v6.2.7 => v6.2.13): Extracting archive
  - Upgrading symfony/form (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/http-client (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/mime (v6.2.10 => v6.2.13): Extracting archive
  - Upgrading symfony/mailer (v6.2.8 => v6.2.12): Extracting archive
  - Upgrading monolog/monolog (3.3.1 => 3.4.0): Extracting archive
  - Upgrading symfony/rate-limiter (v6.2.7 => v6.2.13): Extracting archive
  - Upgrading symfony/serializer (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading symfony/validator (v6.2.11 => v6.2.13): Extracting archive
  - Upgrading twig/cssinliner-extra (v3.6.0 => v3.7.0): Extracting archive
  - Upgrading twig/extra-bundle (v3.6.1 => v3.7.0): Extracting archive
  - Upgrading lorenzo/pinky (1.0.9 => 1.1.0): Extracting archive
  - Upgrading twig/inky-extra (v3.6.0 => v3.7.0): Extracting archive
  - Upgrading symfony/intl (v6.2.10 => v6.2.13): Extracting archive
  - Upgrading twig/intl-extra (v3.6.0 => v3.7.0): Extracting archive
  - Upgrading twig/string-extra (v3.6.0 => v3.7.0): Extracting archive
Generating optimized autoload files
92 packages you are using are looking for funding.
Use the `composer fund` command to find out more!

Run composer recipes at any time to see the status of your Symfony recipes.

Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!   // Clearing the cache for the prod environment with debug false                
!!  
!!  
!!  In Exception.php line 54:
!!                                                                                 
!!    Invalid platform version "" specified. The platform version has to be speci  
!!    fied in the format: "<major_version>.<minor_version>.<patch_version>".       
!!                                                                                 
!!  
!!  cache:clear [--no-warmup] [--no-optional-warmers]
!!  
!!  
Script @auto-scripts was called via post-install-cmd
```

Clearing the cache independently shows the same problem

```
<redacted>$$ bin/console cache:clear --env=prod

 // Clearing the cache for the prod environment with debug false


In Exception.php line 54:

  Invalid platform version "" specified. The platform version has to be specified in the format: "<major_version>.<minor_version>.<patch_version>".


cache:clear [--no-warmup] [--no-optional-warmers]
```

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.0.30

### How do you run Kimai?

Shared-Hosting

### Which PHP version are you using?

8.1

**Comment by kevinpapst** [comment]
Show me your (redacted) database URL. Sounds as if your `serverVersion` is empty.

See https://github.com/kimai/kimai/blob/main/.env.dist#L22 and lines above

**Comment by hkockerbeck** [comment]
The `serverVersion` actually was missing

```
DATABASE_URL=mysql://isatis_db5:<redacted>@localhost/isatis_db5
```

When I add a `serverVersion`, and a `charset` while I'm at it, the cache seems to get cleared:

```
DATABASE_URL=mysql://isatis_db5:<redacted>@localhost/isatis_db5?charset=utf8&serverVersion=mariadb-10.5.19
```

```
<redacted>$ php -d allow_url_fopen=On ~/composer.phar install --no-dev --optimize-autoloader
Installing dependencies from lock file
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating optimized autoload files
92 packages you are using are looking for funding.
Use the `composer fund` command to find out more!

Run composer recipes at any time to see the status of your Symfony recipes.

Executing script cache:clear [OK]
Executing script assets:install [OK]
```
It worked fine without it up to version 2.0.29, though. And when I reverted back to 2.0.29, it still worked without the `serverVersion`. So there's probably a change in version 2.0.30, or in one the bumped dependencies, that caused this changed behaviour.

Thank you for your help :)

**Comment by kevinpapst** [comment]
Seems Doctrine changed its behavior. 
I'll keep this issue open, there will for sure be more users with the same issue.

**Comment by snapet** [comment]
I experienced the same issue. After adding "?charset=utf8&serverVersion=mariadb-10.11.4" the update fell through but when I go my Kimai homepage I now receive:

**_500_**
 **_Something is wrong_**

**_A critical error occurred, please try again. You may have found a software problem, please contact your administrator if the problem persists._** 

Any idea what could be wrong? I cleared the cache already for the production environment.

thanks,
peter

**Comment by hkockerbeck** [comment]
> **_500_** **_Something is wrong_**

The response code 500 is (intentionally) quite vague. In short, it says something like "An internal problem occurred that couldn't be rectified", without any further details. The details should be in your error logs. So you'd probably take a look there.

**Comment by kevinpapst** [comment]
Almost 99.9% of the time when someone has a `500` problem, that someone stopped in the middle of the process and ignored the docs 😈 that say: fix the file permissions (which is also why sometimes people don't even have a log).

**Comment by nemchik** [comment]
Can confirm I had to add `?charset=utf8&serverVersion=mariadb-10.11.4`
It would be nice if there were a generic value that worked the version in case I update my database version without realizing I forgot to update the `DATABASE_URL` variable. Or at least some info on how the version is used, like "it's ok from X to Y range, but needs to be updated if you go to Z"

**Comment by snapet** [comment]
It seems I am in the 0.01% Kevin. The problem was that the environment variable was set both in the .env file and in my Apache server's site file and I had only adapted it in the .env file. For the update commands that works apparently but not when connecting from the web.
I noticed in the logs, which I checked but somehow didn't interpret correctly. It still had the **_Invalid platform version \"\" specified._** error and I forgot I had set the environment variable in the Apache site file also.
Apologies for the noise.

**Comment by kevinpapst** [comment]
That is a rare one @snapet - I think no one posted that before, thanks for sharing 👍

**Comment by divemasterza** [comment]
I can confirm same issue on 2.0.30 (fine up to 2.0.29)
Adding server version and charset fixes it

Q: what will happen when MySQL updates on the server?

**Comment by kevinpapst** [comment]
It's much worse to NOT have that setting instead of wondering what happens when upgrading 😁  that why the version is in the .env since a long time... anyway:

Doctrine (the database abstraction) might generate different SQL depending on a specific version. This is especially important when installing & updating Kimai, as that creates/changes the database structure.

I can't share more details, because I don't know it myself. 
I haven't seen any issue if the minor version is different (which is not a guarantee whatsoever for the future). 
Please upgrade it when you upgrade your database server, especially with a new major version.

**Comment by pbek** [comment]
We need to add the `?charset=utf8&serverVersion=mariadb-10.11.4` part to the examples, like <https://github.com/tobybatch/kimai2/blob/main/compose/docker-compose.apache.prod.yml#L32>. 😅

**Comment by petecooper** [comment]
Field report and tweak for Percona 8.0 + PHP 8.1.

This was my original setting at Kimai 2.0.28:

```
DATABASE_URL=mysql://b309baf5_kimu:{REDACTED}@127.0.0.1:3306/b309baf5_kim?charset=utf8mb4
```

The above generated this error:

```
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!   // Clearing the cache for the prod environment with debug false                
!!  
!!  
!!  In Exception.php line 54:
!!                                                                                 
!!    Invalid platform version "" specified. The platform version has to be speci  
!!    fied in the format: "<major_version>.<minor_version>.<patch_version>".       
!!                                                                                 
!!  
!!  cache:clear [--no-warmup] [--no-optional-warmers]
!!  
!!  
Script @auto-scripts was called via post-install-cmd
```

Appending the `serverVersion` resolved it:

```
DATABASE_URL=mysql://b309baf5_kimu:{REDACTED}@127.0.0.1:3306/b309baf5_kim?charset=utf8mb4&serverVersion=8.0.33-25
```

```
Executing script cache:clear [OK]
Executing script assets:install [OK]


Kimai updates running ...
=========================

[notice] Migrating up to DoctrineMigrations\Version20230819090536
[notice] finished in 666.4ms, used 20M memory, 1 migrations executed, 3 sql queries

                                                                                                                        
 [OK] Successfully migrated to version : DoctrineMigrations\Version20230819090536                                       
                                                                                                                        



 Rebuilding your cache, please be patient ...

 // Clearing the cache for the prod environment with debug false                                                        

                                                                                                                        
 [OK] Cache for the "prod" environment (debug=false) was successfully cleared.                                          
                                                                                                                        


 // Warming up the cache for the prod environment with debug false                                                      

                                                                                                                        
 [OK] Cache for the "prod" environment (debug=false) was successfully warmed.                                           
                                                                                                                        


                                                                                                                        
 [OK] Congratulations! Successfully updated Kimai to version 2.0.31                                                     
                                                                                                                        
```

@divemasterza

>Q: what will happen when MySQL updates on the server?

As a test, I tried `serverVersion=8.0` (i.e. major.minor) on a working 2.0.31 and the cache clearing + warm up appeared to work. I will re-test when Kimai 2.0.next drops.

**Comment by EnMaster** [comment]
Same problem with the docker version 2.0.31
```

Kimai installation running ...
==============================

 [ERROR] Failed to create database: Invalid platform version "" specified. The
         platform version has to be specified in the format:
         "<major_version>.<minor_version>.<patch_version>".

+ /opt/kimai/bin/console -n kimai:update

Kimai updates running ...
=========================

 [ERROR] Failed to validate database.

         Invalid platform version "" specified. The platform version has to be
         specified in the format:
         "<major_version>.<minor_version>.<patch_version>".
```

After adding `?charset=utf8&serverVersion=mariadb-10.1.48` to my `DATABASE_URL`

```
Kimai installation running ...
==============================

Database `kimai` for connection named default already exists. Skipped.
[notice] Migrating up to DoctrineMigrations\Version20230819090536
[notice] finished in 1334.3ms, used 74.5M memory, 1 migrations executed, 3 sql queries

 [OK] Successfully migrated to version :
      DoctrineMigrations\Version20230819090536
```

Now all works :D

**Comment by petecooper** [comment]
https://github.com/kimai/kimai/issues/4246#issuecomment-1692019630

>As a test, I tried serverVersion=8.0 (i.e. major.minor) on a working 2.0.31 and the cache clearing + warm up appeared to work. I will re-test when Kimai 2.0.next drops.

Successful upgrade from 2.0.31 to 2.0.32 with `serverVersion=8.0` (i.e. major.minor, no vendor) in `.env`:

```
remote: Enumerating objects: 733, done.
remote: Counting objects: 100% (336/336), done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 87 (delta 77), reused 62 (delta 57), pack-reused 0
Unpacking objects: 100% (87/87), 134.69 KiB | 1.45 MiB/s, done.
From https://github.com/kevinpapst/kimai2
 * [new tag]           2.0.32     -> 2.0.32
Previous HEAD position was 636422e3 show global form errors - fixes #4233
HEAD is now at 31348da4 Release 2.0.32 (#4256)
Installing dependencies from lock file
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating optimized autoload files
92 packages you are using are looking for funding.
Use the `composer fund` command to find out more!

Run composer recipes at any time to see the status of your Symfony recipes.

Executing script cache:clear [OK]
Executing script assets:install [OK]


Kimai updates running ...
=========================


                                                                                                                        
 [OK] Already at the latest version ("DoctrineMigrations\Version20230819090536")                                        
                                                                                                                        


 Rebuilding your cache, please be patient ...

 // Clearing the cache for the prod environment with debug false                                                        

                                                                                                                        
 [OK] Cache for the "prod" environment (debug=false) was successfully cleared.                                          
                                                                                                                        


 // Warming up the cache for the prod environment with debug false                                                      

                                                                                                                        
 [OK] Cache for the "prod" environment (debug=false) was successfully warmed.                                           
                                                                                                                        


                                                                                                                        
 [OK] Congratulations! Successfully updated Kimai to version 2.0.32                                                     
                                                                                                                        
```

**Comment by kevinpapst** [comment]
Docs have been updated. Closing.
